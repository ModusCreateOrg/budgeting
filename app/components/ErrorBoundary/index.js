// @flow
import * as React from 'react';

type FallbackComponentProps = {
  error: Error,
  info: Object,
};

type ErrorBoundaryProps = {
  children: React.Node,
  fallbackComponent: React.ComponentType<FallbackComponentProps>,
  onError: ?(error: Error, info: Object) => void,
};

type ErrorBoundaryState = {
  error: ?Error,
  info: ?Object,
};

class ErrorBoundary extends React.Component<ErrorBoundaryProps, ErrorBoundaryState> {
  static defaultProps = {
    onError: null,
  };

  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { error: null, info: null };
  }

  componentDidCatch(error: Error, info: Object) {
    // Catch errors in any components below and re-render with error message
    this.setState({
      error: error,
      info: info,
    });

    // A callback can be used to log error messages
    if (this.props.onError) {
      this.props.onError(error, info);
    }
  }

  render() {
    const { fallbackComponent } = this.props;
    const { error, info } = this.state;

    if (error && info) {
      // Render error component
      return React.createElement(fallbackComponent, { error, info });
    }

    // Normally, just render children
    return this.props.children;
  }
}

export default ErrorBoundary;
