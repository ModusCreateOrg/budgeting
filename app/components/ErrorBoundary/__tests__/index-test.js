import React from 'react';
import renderer from 'react-test-renderer';
import ErrorBoundary from '../';

const FailingComponent = () => {
  throw new Error('I crashed');
};

const WorkingComponent = () => <div>working</div>;

const FallbackComponent = () => <div>fallback</div>;

beforeAll(() => {
  // A console.error is expected here, saying that ErrorBoundary will handle an error.
  // We mock this to see a clean output in the test runs.
  console.error = jest.fn();
});

it('renders a working component', () => {
  const tree = renderer
    .create(
      <ErrorBoundary fallbackComponent={FallbackComponent}>
        <WorkingComponent />
      </ErrorBoundary>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders a fallback component when the children component fails', () => {
  const tree = renderer
    .create(
      <ErrorBoundary fallbackComponent={FallbackComponent}>
        <FailingComponent />
      </ErrorBoundary>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('calls an error handler when the children component fails', () => {
  const handleError = jest.fn();

  renderer.create(
    <ErrorBoundary fallbackComponent={FallbackComponent} onError={handleError}>
      <FailingComponent />
    </ErrorBoundary>
  );

  expect(handleError).toHaveBeenCalled();
});
