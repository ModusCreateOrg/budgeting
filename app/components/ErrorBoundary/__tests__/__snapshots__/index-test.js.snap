// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a fallback component when the children component fails 1`] = `
<div>
  fallback
</div>
`;

exports[`renders a working component 1`] = `
<div>
  working
</div>
`;
